File: backend/src/utils/response.ts
import { Response } from 'express';

export function success<T>(res: Response, data: T, meta?: { page: number; limit: number; total: number }) {
  res.json({ success: true, data, ...(meta && { meta }) });
}

export function created<T>(res: Response, data: T) {
  res.status(201).json({ success: true, data });
}

export function error(res: Response, status: number, code: string, message: string, details?: any) {
  res.status(status).json({ success: false, error: { code, message, ...(details && { details }) } });
}

File: backend-py/app/core/response.py
from __future__ import annotations

from typing import Any

from fastapi.encoders import jsonable_encoder
from fastapi.responses import JSONResponse


def success(data: Any, meta: dict | None = None) -> JSONResponse:
    body: dict = {"success": True, "data": jsonable_encoder(data)}
    if meta:
        body["meta"] = jsonable_encoder(meta)
    return JSONResponse(content=body)


def created(data: Any) -> JSONResponse:
    return JSONResponse(status_code=201, content={"success": True, "data": jsonable_encoder(data)})


def error(status: int, code: str, message: str, details: Any = None) -> JSONResponse:
    err: dict = {"code": code, "message": message}
    if details is not None:
        err["details"] = jsonable_encoder(details)
    return JSONResponse(status_code=status, content={"success": False, "error": err})
